### PR TITLE
WIP: Blob streams

### PR DESCRIPTION
**Pull request by Dakror**

### CODE REVIEW

File: webdb/src/webdb_bindings.ts [hunk 1]
```diff
@@ -24,6 +32,25 @@ function copyFlatbuffer(buffer: Uint8Array): flatbuffers.ByteBuffer {
...
+    public buffer: Uint8Array;
+    public position: number;
+
+    public constructor(buffer: Uint8Array) {
+        this.buffer = buffer;
+        this.position = 0;
+    }
+}
+
+// As a global function because when passing the object to the webworker it turns into a POJO (?)
```
Comment (ankoh) [inline webdb/src/webdb_bindings.ts]: Yes, that's a problem. Maybe just turn the blobstream into an interface?

File: webdb/src/async_webdb.ts [hunk 2]
```diff
@@ -214,6 +216,15 @@ export class AsyncWebDB {
         await this.postTask(task);
     }
 
+    /// Ping the worker thread
```
Comment (ankoh) [inline webdb/src/async_webdb.ts]: Comment

File: webdb/src/worker_node.ts [hunk 3]
```diff
@@ -2,19 +2,29 @@
...
         globalThis.postMessage(response, transfer);
     }
 
     /// Instantiate the wasm module
     protected async open(path: string | null): Promise<WebDBBindings> {
-        const bindings = new WebDB(this, {}, path);
+        const bindings = new WebDB(
+            this,
+            {
+                dashql_blob_stream_underflow: function (blobId: number, buf: number, size: number): number {
```
Comment (ankoh) [inline webdb/src/worker_node.ts]: Ok here for now but we should probably just move this to a dedicated node_runtime file or so

**Review by ankoh — CHANGES_REQUESTED** [review]
Very nice!

File: webdb/src/worker_node.ts (hunk 3 above)
Comment (Dakror) [reply webdb/src/worker_node.ts]: The problem here is the circular referencing of runtime and bindings. if i move the runtime to a new file i somehow need to give it the bindings, but also the other way around. whats the best solution there?

Comment (ankoh) [reply webdb/src/worker_node.ts]: Ah I see. Then lets leave it like that at the moment. I'd probably just move the blob map into the runtime as well and then pass the runtime to the bindings on init.

File: benchmarks/src/iterator_benchmark.ts [hunk 4]
```diff
@@ -526,7 +526,7 @@ function main(db: webdb.WebDB) {
 }
 
 const logger = new webdb.VoidLogger();
-const db = new webdb.WebDB(logger, {}, wasmPath);
+const db = new webdb.WebDB(logger, new webdb.DefaultWebDBRuntime(), wasmPath);
```
Comment (ankoh) [inline benchmarks/src/iterator_benchmark.ts]: static variable now, no need for new

File: benchmarks/src/proxy_benchmark.ts [hunk 5]
```diff
@@ -69,7 +69,7 @@ function main(db: webdb.WebDB) {
 }
 
 const logger = new webdb.VoidLogger();
-const db = new webdb.WebDB(logger, {}, wasmPath);
+const db = new webdb.WebDB(logger, new webdb.DefaultWebDBRuntime(), wasmPath);
```
Comment (ankoh) [inline benchmarks/src/proxy_benchmark.ts]: static variable now, no need for new

File: webdb/package.json [hunk 6]
```diff
@@ -30,7 +30,7 @@
         "fsevents": "*"
     },
     "scripts": {
-        "build": "webpack --mode=production && node ./gendts.js",
+        "build": "webpack --mode=development && node ./gendts.js",
```
Comment (ankoh) [inline webdb/package.json]: production

File: webdb/test/webdb_bindings.test.ts [hunk 7]
```diff
@@ -7,15 +7,15 @@ var conn: webdb.WebDBConnection;
 const logger = new webdb.ConsoleLogger();
 
 beforeAll(async () => {
-    db = new webdb.WebDB(logger, {}, path.resolve(__dirname, "../src/webdb_wasm.wasm"));
+    db = new webdb.WebDB(logger, new webdb.DefaultWebDBRuntime(), path.resolve(__dirname, '../src/webdb_wasm.wasm'));
```
Comment (ankoh) [inline webdb/test/webdb_bindings.test.ts]: static pojo now

File: webdb/test/proxy.test.ts [hunk 8]
```diff
@@ -8,7 +8,7 @@ const testRows = 3000;
 const logger = new webdb.ConsoleLogger();
 
 beforeAll(async () => {
-    db = new webdb.WebDB(logger, {}, path.resolve(__dirname, '../src/webdb_wasm.wasm'));
+    db = new webdb.WebDB(logger, new webdb.DefaultWebDBRuntime(), path.resolve(__dirname, '../src/webdb_wasm.wasm'));
```
Comment (ankoh) [inline webdb/test/proxy.test.ts]: here as well

**Review by ankoh — CHANGES_REQUESTED** [review]

**Review by ankoh — COMMENTED** [review]
Awesome :+1:

**Pull request merged**